src/compose/images.ts

```diff
@@ -398,35 +434,75 @@ async function getImagesForCleanup(): Promise<string[]> {
 		.value();
 }
 
-export async function inspectByName(
-	imageName: string,
-): Promise<Docker.ImageInspectInfo> {
-	try {
-		const image = await docker.getImage(imageName);
-		return await image.inspect();
-	} catch (e) {
-		if (NotFoundError(e)) {
-			const digest = imageName.split('@')[1];
-			let imagesFromDb: Image[];
-			if (digest != null) {
-				imagesFromDb = await db
-					.models('image')
-					.where('name', 'like', `%@${digest}`);
-			} else {
-				imagesFromDb = await db
-					.models('image')
-					.where({ name: imageName })
-					.select();
-			}
+export async function inspectByName(imageName: string) {
+	// Fail fast
+	assert(!!imageName, 'image name to inspect cannot be null');
 
-			for (const image of imagesFromDb) {
-				if (image.dockerImageId != null) {
-					return await docker.getImage(image.dockerImageId).inspect();
-				}
-			}
-		}
-		throw e;
-	}
+	// Get image by the full image URI. This will only work for regular pulls
+	// and old style images `repo:tag`
+	const inspectByURI = async () => await docker.getImage(imageName).inspect();
+
+	const {
+		registry,
+		imageName: name,
+		tagName,
+		digest,
+	} = dockerUtils.getRegistryAndName(imageName);
+
+	// Look for an image in the engine with registry/image as reference (tag)
+	// for images with deltas this should return unless there is some inconsistency
+	// and the tag was deleted
+	const repo = [registry, name].filter((s) => !!s).join('/');
+	const reference = [repo, tagName].filter((s) => !!s).join(':');
+
+	const inspectByReference = async () =>
+		await docker
+			.listImages({
+				digests: true,
+				filters: { reference: [reference] },
+			})
+			.then(([img]) =>
+				!!img
+					? docker.getImage(img.Id).inspect()
+					: Promise.reject(
+							new StatusError(
+								404,
+								`Failed to find an image matching ${imageName}`,
+							),
+					  ),
+			);
+
+	// Otherwise look in the database for an image with same digest or same name and
+	// get the dockerImageId from there. If this fails the image may still be on the
+	// engine but we need to re-trigger fetch and let the engine tell us if the
+	// image data is there
+	const inspectByDigest = async () =>
+		await db
+			.models('image')
+			.where('name', 'like', `%${digest}`)
+			.orWhere({ name: imageName }) // Default to looking for the full image name
+			.select()
+			.then((images) =>
+				images.filter((img: Image) => img.dockerImageId !== null),
+			)
+			// Assume that all db entries will point to the same dockerImageId, so use
+			// the first one. If this assumption is false, there is a bug with cleanup
+			.then(([img]) =>
+				!!img
+					? docker.getImage(img.dockerImageId).inspect()
+					: Promise.reject(
+							new StatusError(
+								404,
+								`Failed to find an image matching ${imageName}`,
+							),
+					  ),
+			);
+
+	// Convert the promise array into a promise chain
+	return await [inspectByURI, inspectByReference, inspectByDigest].reduce(
```

Comment:
Didn't see the point in declaring them outside really, since they are not being used elsewhere. Doing it this way means we only need to perform the `assert` once. 

If we ever need to use them individually we can always refactor, if you feel strongly about it I can make the change but I don't think is that much of an improvement.

Reply:
I think declaring them outside the function would improve the readability of that function with no cost